RaftMerge/RaftMerge.tla

```diff
@@ -0,0 +1,297 @@
+HandleAppendEntriesReply(i, j, r, m) ==
+  /\ raft' = [raft EXCEPT ![i][r].match_index = [@ EXCEPT ![j] = Max({@, m.match_index})]]
+  /\ Discard(m)
+  /\ UNCHANGED <<region, client_vars>>
+
+\* Poll a message from message set, and invoke the corresponding RPC handler.
+Receive(m) ==
+  LET
+    i == m.dest
+    j == m.source
+    r == m.region
+  IN
+    \/ /\ m.type = AppendEntriesRequest
+       /\ HandleAppendEntriesRequest(i, j, r, m)
+    \/ /\ m.type = AppendEntriesReply
+       /\ HandleAppendEntriesReply(i, j, r, m)
+
+\* Leader i of region r receives a client request to append a log.
+ClientRequest(i, r, log) ==
+  /\ raft[i][r].is_leader
+  /\ client_requests_index < MaxClientRequests
+  /\ LET
+       new_logs == Append(raft[i][r].logs, log)
+       new_match_index == [raft[i][r].match_index EXCEPT ![i] = @ + 1]
+     IN
+       /\ raft' = [raft EXCEPT ![i][r].logs = new_logs,
+                               ![i][r].match_index = new_match_index]
+       /\ client_requests_index' = client_requests_index + 1
+  /\ UNCHANGED <<messages, region>>
+
+-------------------------------------------------------------------------------
+
+\* Return TRUE if there is a log applicable to the state machine.
+LogAppliable(i, r) ==
+  raft[i][r].apply_index < raft[i][r].commit_index
+
+\* Apply Raft logs to make apply_index catch up with commit_index.
+\* This simply increases apply_index.
+ApplyLog(i, r) ==
+  LET
+    next_index == raft[i][r].apply_index + 1
+  IN
+    /\ LogAppliable(i, r)
+    /\ raft' = [raft EXCEPT ![i][r].apply_index = next_index]
+    /\ UNCHANGED <<messages, region, client_vars>>
+
+-------------------------------------------------------------------------------
+
+\* Specification of Raft merge.
+Next ==
+  \* Raft actions.
+  \/ \E i,j \in Store : \E r \in Region : AppendEntries(i, j, r)
+  \/ \E i \in Store : \E r \in Region : AdvanceCommitIndex(i, r)
+  \/ \E m \in messages : Receive(m)
+  \/ \E i \in Store : \E r \in Region : ApplyLog(i, r)
+
+  \* External client can send requests to region leader.
+  \/ \E i \in Store : \E r \in Region :
+        ClientRequest(i, r, Log)
+
+Init ==
+  /\ messages = {}
+  /\ LET
+       \* Mark region r on store i as leader.
+       MarkLeader(rafts, i, r) == [rafts EXCEPT ![i] = [@ EXCEPT ![r].is_leader = TRUE]]
+       no_leader_raft == [i \in Store |->
+                           [r \in Region |->
+                             [is_leader    |-> FALSE,
+                              logs         |-> << >>,
+                              commit_index |-> 0,
+                              apply_index  |-> 0,
+                              match_index  |-> [j \in Store |-> 0]
+                             ]
+                           ]
+                         ]
+     IN
+       raft = MarkLeader(MarkLeader(no_leader_raft, LeaderA, RegionA),
+                         LeaderB,
+                         RegionB)
+  /\ region = TRUE
```

Comment:
why TRUE here? where to use region?

Reply:
This is a variable will be used in later PR.